qmp: Fix race causing events to be sent during negotiation

As per the QMP spec, asynchronous messages should not be sent during
negotiation.

The event sending code checks if the monitor is in the negotiation phase
by checking for mon->commands != qmp_cap_negotiation_commands. However,
events may be incorrectly sent from the point the connection is opened
to when monitor_qmp_event() sets the negotiation phase.

Ensure it is always in the negotiation phase when a connection is opened
by initializing it during monitor init and close.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
diff --git a/monitor/qmp.c b/monitor/qmp.c
index b67a8e7d1f63..19e2edfd1cf8 100644
--- a/monitor/qmp.c
+++ b/monitor/qmp.c
@@ -344,7 +344,6 @@ static void monitor_qmp_event(void *opaque, int event)
 
     switch (event) {
     case CHR_EVENT_OPENED:
-        mon->commands = &qmp_cap_negotiation_commands;
         monitor_qmp_caps_reset(mon);
         data = qmp_greeting(mon);
         qmp_send_response(mon, data);
@@ -364,6 +363,7 @@ static void monitor_qmp_event(void *opaque, int event)
                                  mon, NULL);
         mon_refcount--;
         monitor_fdsets_cleanup();
+        mon->commands = &qmp_cap_negotiation_commands;
         break;
     }
 }
@@ -398,6 +398,7 @@ void monitor_init_qmp(Chardev *chr, bool pretty)
     monitor_data_init(&mon->common, true, false,
                       qemu_chr_has_feature(chr, QEMU_CHAR_FEATURE_GCONTEXT));
 
+    mon->commands = &qmp_cap_negotiation_commands;
     mon->pretty = pretty;
 
     qemu_mutex_init(&mon->qmp_queue_lock);
